File: emulator/c/hw_she.c
// Sample Crypt primitives based on AES hash, CMAC and KDF for MARS
// Implements algorithms specified in AUTOSAR Secure Hardware Extensions (SHE)
// Author: Tom Brostrom

// AES implementation from:
// Tiny AES, https://github.com/kokke/tiny-AES-c

#include "../aes/aes.h" // Tiny AES

#include <string.h>   // for memset, memcpy
#include <stdio.h>
#include <stdbool.h>

#include "hw_she.h"
#include "mars_internal.h"

static void hexout(const char *msg, const void *buf, size_t len)
{
typeof(len) i;
    if (msg)
        printf("%s: ", msg);
    for (i=0; i<len; i++)
        printf("%02x", ((uint8_t *)buf)[i]);
    printf("\n");
}

// Miyaguchi–Preneel (MP) compression
// https://en.wikipedia.org/wiki/One-way_compression_function#Miyaguchi–Preneel
// Apply MP round to each block in message M
// Any trailing incomplete block is ignored
static void mp_comp(const uint8_t * M, size_t n, uint8_t * H)
{
size_t j, b = n >> 4;  // number of full blocks to process, n / 16
struct AES_ctx ctx;
uint8_t EkM[16];

    while (b--)
        {
        AES_init_ctx(&ctx, H);
        memcpy(&EkM, M, sizeof(EkM));
        AES_ECB_encrypt(&ctx, EkM);
        for (j=0; j<16; j++)
            H[j] ^= EkM[j] ^ M[j];
        M += 16;
        }
}

void CryptHashInit(profile_shc_t *hctx)
{
    memset(&hctx->H, 0, 16);
    hctx->total = 0;
    hctx->len = 0;
}

#define MIN(X, Y) (((X) < (Y)) ? (X) : (Y))

// Hash blocks from previous partial block (if any) and msg
// Bytes from trailing incomplete block in msg are copied to blk
void CryptHashUpdate(profile_shc_t *hctx, const void * msg, size_t n)
{
size_t pn;  // number of bytes to append to partial block
    hctx->total += n;
    if (hctx->len)   // any bytes from partial block?
        {               // try to fill blk from msg
        pn = MIN(16-hctx->len, n);
        memcpy(&hctx->blk[hctx->len], msg, pn);
        hctx->len += pn;
        msg += pn;
        n -= pn;
        if (hctx->len == 16)
            {           // blk is full, compress it
            mp_comp(hctx->blk, 16, hctx->H);
            hctx->len = 0;
            }
        }
     if (n)     // additional bytes in msg
        {       // len should be 0 at this point
        mp_comp(msg, n, hctx->H);
        pn = n & 0xf;   // n % 16, bytes from trailing partial block
        msg += n - pn;  // move trailing pn bytes to blk
        hctx->len = pn;
        memcpy(&hctx->blk, msg, pn);
        }
}

// pad and do final compress(es), return digest
void CryptHashFinal(profile_shc_t *hctx, void *dig)
{
uint8_t i, len = hctx->len;
uint8_t np = 16 - 1 - len; // # of bytes free for padding, excludes termination byte
size_t bits = hctx->total << 3; // total * 8

    // add termination bit in 1 byte
    // always room for 1 byte - otherwise wouldn't be partial
    hctx->blk[len++] = 0x80;

    // If no room for 5-byte length, just 0 pad and compress
    if (np < 5) {
        memset(&hctx->blk[len], 0, np);
        mp_comp(hctx->blk, 16, hctx->H);
        len = 0;
        np = 16;
    }

    // 0 pad, append length, compress
    memset(&hctx->blk[len], 0, np-5);
    for (i=5; i-->0; )      // i = 4..0
        hctx->blk[15-i] = bits >> (i<<3);   // append total bits, big endian in 5 bytes
    mp_comp(hctx->blk, 16, hctx->H);

    memcpy(dig, hctx->H, sizeof(hctx->H));
}


// Left shift byte array by 1 bit
// Return the final carry bit
static uint8_t ls1(uint8_t *a, size_t n)
{
uint8_t C2, C = 0;

    while (n--)
        {
        C2 = a[n] >> 7;
        a[n] = (a[n] << 1) | C;
        C = C2;
        }
    return C;
}

// XOR Byte Arrays, x = x ^ y
static void xba(uint8_t * x, const uint8_t *y, size_t n)
{
    while (n--)
        x[n] ^= y[n];
}

// Simplified CMAC algorithm for a single block.
void CryptSign(void *mac, const void *key, const void *blk)
{
struct AES_ctx ctx;

    AES_init_ctx(&ctx, key);
    memset(mac, 0, 16);
    AES_ECB_encrypt(&ctx, mac);
    if (ls1(mac,16))
        ((uint8_t *)mac)[15] ^= 0x87;
    xba(mac, blk, 16);
    AES_ECB_encrypt(&ctx, mac);
}


bool CryptVerify(const void *key, const void *dig, const void *sig)
{
uint8_t mac[16];
    CryptSign(mac, key, dig);
    return memcmp(mac, sig, 16) == 0;
}

// This KDF is adapted from the AUTOSAR version where
// label and context replace "SHE".
void CryptSkdf(void * key, const void * parent, char label, const void * ctx, uint16_t ctxlen)
{
she_hctx_t hctx;
    CryptHashInit(&hctx);
    CryptHashUpdate(&hctx, parent, 16);
    CryptHashUpdate(&hctx, "\x01\x01", 2);
    CryptHashUpdate(&hctx, &label, 1);
    CryptHashUpdate(&hctx, ctx, ctxlen);
    CryptHashUpdate(&hctx, "", 1);      // addr of null byte
    CryptHashFinal(&hctx, key);
}

void CryptXkdf(void * key, const void * parent, char label, const void * ctx, uint16_t ctxlen)
{
    CryptSkdf(key, parent, label, ctx, ctxlen);
}


// These tests are from the AUTOSAR SHE spec, 4.13 Examples and Test Vectors
// See https://www.autosar.org/fileadmin/standards/foundation/22-11/AUTOSAR_TR_SecureHardwareExtensions.pdf
bool CryptSelfTest(bool fullTest)
{
size_t i;
uint8_t key[16], out[16];
uint8_t key1[16] =  { 0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
                      0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c };
uint8_t msg1[16] =  { 0x6b, 0xc1, 0xbe, 0xe2, 0x2e, 0x40, 0x9f, 0x96,
                      0xe9, 0x3d, 0x7e, 0x11, 0x73, 0x93, 0x17, 0x2a };
uint8_t exp1[16] =  { 0x07, 0x0a, 0x16, 0xb4, 0x6b, 0x4d, 0x41, 0x44,
                      0xf7, 0x9b, 0xdd, 0x9d, 0xd0, 0x4a, 0x28, 0x7c };

uint8_t exp2[16] =  { 0x11, 0x8a, 0x46, 0x44, 0x7a, 0x77, 0x0d, 0x87,
                      0x82, 0x8a, 0x69, 0xc2, 0x22, 0xe2, 0xd1, 0x7e };

// TEST 1 for CMAC, from spec 4.13.2.3, example 1
    CryptSign(out, key1, msg1);
    if (memcmp(out, exp1, 16) != 0)
        return false;

// TEST2 for KDF, from spec 4.13.2.5
    for (i=0; i<sizeof(key); i++)
        key[i] = i;
    CryptSkdf(out, key, 'S', "HE", 2);
    if (memcmp(out, exp2, 16) != 0)
        return false;

    return true;
}

File: emulator/python/hw_she.py
#!/usr/bin/env python3

# Sample Crypt primitives based on AES hash, CMAC and KDF for MARS
# Implements algorithms specified in AUTOSAR Secure Hardware Extensions (SHE)
# Requires pycryptodome
# Author: Tom Brostrom, CPVI

import Crypto.Cipher.AES as aes
import aes_mphash as hashmod

len_digest = hashmod.digest_size
len_sign = 16
len_skey = 16
len_akey = 0
alg_hash = 0x84  # TPM_ALG_? AES_MP
alg_sign = 0x3F # TPM_ALG_CMAC
alg_skdf = 0x86 # unused
alg_akdf = 0    # TPM_ALG_ERROR

# convert int i to big endian array of n bytes
def int2bebar(i, n):
    return bytes([i>>(j<<3) & 0xff for j in reversed(range(n))])

# SHE only supports ECB (default) and CBC modes
# See tests below for examples
def cipher_ecb(key): return aes.new(key, aes.MODE_ECB)
def cipher_cbc(key, iv): return aes.new(key, aes.MODE_CBC, iv=iv)

# Left shift bytearray by 1 bit. Return MSB.
def ls1(a):
    C = 0
    for i in reversed(range(len(a))):
        b = (a[i] << 1) | C
        a[i] = b & 0xff 
        C = b >> 8
    return C

# XOR Byte Arrays # compute x = x ^ y
def xba(x, y):
    assert len(x) == len(y)
    for i in range(len(x)):
        x[i] ^= y[i]

# Simplified CMAC algorithm for a single block.
def cmac1(K, blk):
    assert len(blk) == 16
    E = cipher_ecb(K).encrypt
    mtk = bytearray(E(bytes(16)))       # create the Tweak sub-key k0
    if ls1(mtk): mtk[15] ^= 0x87        # turn k0 into k1
    xba(mtk, blk)                       # create Mn' from k1 ^ blk
    return E(bytes(mtk))                # then encrypt Mn'

# Full CMAC algorithm.
# Unused here.
# Verify with: https://artjomb.github.io/cryptojs-extension/
def cmac(K, msg):
    # print('key =', K.hex())
    E = cipher_ecb(K).encrypt
    mtk = bytearray(E(bytes(16)))       # create the Tweak sub-key k0
    if ls1(mtk): mtk[15] ^= 0x87        # turn k0 into k1

    f = len(msg) >> 4   # full blocks in msg
    r = len(msg) & 0xf  # remainder bytes in last block of msg
    # print('msg =', msg.hex(), " f =", f, ' r =', r)
    if r or not f:                      # partial block, or empty msg
        if ls1(mtk): mtk[15] ^= 0x87    # turn k1 into k2
        last = msg[-r:] + b'\x80' + bytes(15-r)
    else:
        f -= 1
        last = msg[-16:]

    # XOR the tweak sub-key (k1 or k2 in mtk) with last block
    xba(mtk, last)                      # Create Mn': mtk ^= last

    # process all but the last block
    V = bytearray(16)
    for i in range(f):
        xba(V, msg[i<<4:(i+1)<<4])      # V ^= Mi
        V = bytearray(E(bytes(V)))      # V = E(V)

    # process the last tweaked block Mn' in mtk
    xba(V, mtk)                         # V ^= Mn'
    V = E(bytes(V))                     # V = E(V)
    return V

def CryptHash(msg):
    return hashmod.new(msg).digest()

CryptSign = cmac1

def CryptVerify(key, dig, sig):
    return sig == CryptSign(key, dig)

# This KDF is adapted from the AUTOSAR version where
# label and context replace "SHE".
def CryptSkdf(key, label, ctx):
    assert len(label) == 1
    return CryptHash(key + b'\x01\x01' + label + ctx + b'\x00')

CryptAkdf = None

# These tests are from the AUTOSAR SHE spec, 4.13 Examples and Test Vectors
# See https://www.autosar.org/fileadmin/standards/foundation/22-11/AUTOSAR_TR_SecureHardwareExtensions.pdf
def CryptSelfTest(fullTest):

    # ECB TESTS, from spec 4.13.1
    key = bytes.fromhex('000102030405060708090a0b0c0d0e0f')
    c = cipher_ecb(key)
    pt  = bytes.fromhex('00112233445566778899aabbccddeeff')
    exp = bytes.fromhex('69c4e0d86a7b0430d8cdb78070b4c55a')
    out = c.encrypt(pt)
    assert out == exp
    out = c.decrypt(out)
    assert out == pt

    # CMAC TEST, from spec 4.13.2.3, example 1
    # Same test as RFC 4493, Example 2
    key = bytes.fromhex('2b7e151628aed2a6abf7158809cf4f3c')
    msg = bytes.fromhex(
        '6bc1bee22e409f96e93d7e117393172aae2d8a571e03ac9c9eb76fac45af8e51')
    exp = bytes.fromhex('070a16b46b4d4144f79bdd9dd04a287c')
    out = CryptSign(key, msg[:16])
    assert out == exp

    # HASH TEST, from spec 4.13.2.4
    exp = bytes.fromhex('c7277a0dc1fb853b5f4d9cbd26be40c6')
    out = CryptHash(msg)
    assert out == exp

    # HASH UPDATE TEST
    hobj = hashmod.new()
    hobj.update(msg[0:4])
    hobj.update(msg[4:21])
    hobj.update(msg[21:32])
    out = hobj.digest()
    assert out == exp
 
    # KDF TEST, from spec 4.13.2.5
    key = bytes.fromhex('000102030405060708090a0b0c0d0e0f')
    exp = bytes.fromhex('118a46447a770d87828a69c222e2d17e')
    out = CryptSkdf(key, b'S', b'HE')
    assert out == exp

    return True


if __name__ == "__main__":

    print('SELF TEST:', 'pass' if CryptSelfTest(True) else 'FAIL')

